bundle/src/main/java/com/adobe/acs/commons/remoteassets/impl/RemoteAssetDecorator.java

```diff
@@ -0,0 +1,175 @@
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceDecorator;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.jcr.base.util.AccessControlUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.servlet.http.HttpServletRequest;
+import java.util.Calendar;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+/**
+ * ResourceDecorator that instruments remote assets to sync binaries as needed.
+ * This "decorator" is used to detect the first time a "remote" asset is
+ * referenced by the system and sync that asset from the remote server to
+ * make it now a "true" asset.
+ */
+@Component(
+        label = "ACS AEM Commons - Remote Assets - Asset Resource Decorator",
+        description = "Captures a request for a remote asset so that the binary can be sync'd to the current server making it a true local asset",
+        policy = ConfigurationPolicy.REQUIRE
+)
+@Service
+public class RemoteAssetDecorator implements ResourceDecorator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RemoteAssetDecorator.class);
+
+    /**
+     * This set stores resource paths for remote assets that are in the process
+     * of being sync'd from the remote server.  This prevents an infinite loop
+     * when the RemoteAssetSync service fetches the asset in order to update it.
+     */
+    private static Set<String> remoteResourcesSyncing = new ConcurrentSkipListSet<>();
+
+    @Reference
+    private RemoteAssetsBinarySync assetSync;
+
+    @Reference
+    private RemoteAssetsConfig config;
+
+    /**
+     * When resolving a remote asset, first sync the asset from the remote server.
+     * @param resource The resource being resolved.
+     * @return The current resource.  If the resource is a "remote" asset, it will
+     * first be converted to a true local AEM asset by sync'ing in the rendition
+     * binaries from the remote server.
+     */
+    @Override
+    public Resource decorate(final Resource resource) {
+        try {
+            if (!this.accepts(resource)) {
+                return resource;
+            }
+        } catch (Exception e) {
+            // Logging at debug level b/c if this happens it could represent a ton of logging
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Failed binary sync check for remote asset: {} - {}", resource.getPath(), e.getMessage());
+            }
+            return resource;
+        }
+
+        Resource ret = resource;
+        try {
+            remoteResourcesSyncing.add(resource.getPath());
+            LOG.info("Sync'ing remote asset binaries: {}", resource.getPath());
+            ret = this.assetSync.syncAsset(resource);
+        } catch (Exception e) {
+            LOG.error("Failed to sync binaries for remote asset: {} - {}", resource.getPath(), e.getMessage());
```

Comment:
Same mentality as above - trying to reduce a done of repeated log clutter if possible.  But I can see that this could potentially make debugging more tricky.  Thoughts?